src/linux/module_reader.rs

```diff
@@ -0,0 +1,433 @@
+    for header in section_headers {
+        let sh_name = header.sh_name as u64;
+        if sh_name >= strtab_section_header.sh_size {
+            log::warn!("invalid sh_name offset");
+            continue;
+        }
+        if sh_name + name.len() as u64 >= strtab_section_header.sh_size {
+            // This can't be a match.
+            continue;
+        }
+        let n = read(
+            module_memory,
+            strtab_section_header.sh_offset + sh_name,
+            name.len() as u64,
+        )?;
+        if name == &*n {
+            return Ok(Some(header));
+        }
+    }
+    Ok(None)
+}
+
+/// Types which can be read from an `impl ModuleMemory`.
+pub trait ReadFromModule: Sized {
+    fn read_from_module(module_memory: impl ModuleMemory) -> Result<Self, Error>;
+}
+
+/// The module build id.
+#[derive(Default, Clone, Debug)]
+pub struct BuildId(pub Vec<u8>);
+
+impl ReadFromModule for BuildId {
+    fn read_from_module(module_memory: impl ModuleMemory) -> Result<Self, Error> {
+        let reader = ModuleReader::new(module_memory)?;
+        let program_headers = match reader.build_id_from_program_headers() {
+            Ok(v) => return Ok(BuildId(v)),
+            Err(e) => Box::new(e),
+        };
+        let section = match reader.build_id_from_section() {
+            Ok(v) => return Ok(BuildId(v)),
+            Err(e) => Box::new(e),
+        };
+        let generated = match reader.build_id_generate_from_text() {
+            Ok(v) => return Ok(BuildId(v)),
+            Err(e) => Box::new(e),
+        };
+        Err(Error::NoBuildId {
+            program_headers,
+            section,
+            generated,
+        })
+    }
+}
+
+/// The module SONAME.
+#[derive(Default, Clone, Debug)]
+pub struct SoName(pub String);
+
+impl ReadFromModule for SoName {
+    fn read_from_module(module_memory: impl ModuleMemory) -> Result<Self, Error> {
+        ModuleReader::new(module_memory)
+            .and_then(|r| r.soname())
+            .map(SoName)
+    }
+}
+
+pub struct ModuleReader<T> {
+    module_memory: T,
+    header: elf::Header,
+    context: Ctx,
+}
+
+impl<T: ModuleMemory> ModuleReader<T> {
+    pub fn new(module_memory: T) -> Result<Self, Error> {
+        // We could use `Ctx::default()` (which defaults to the native system), however to be extra
+        // permissive we'll just use a 64-bit ("Big") context which would result in the largest
+        // possible header size.
+        let header_size = elf::Header::size(Ctx::new(Container::Big, Endian::default()));
+        let header_data = read(&module_memory, 0, header_size as u64)?;
+        let header = elf::Elf::parse_header(&header_data)?;
+        let context = Ctx::new(header.container()?, header.endianness()?);
+        Ok(ModuleReader {
+            module_memory,
+            header,
+            context,
+        })
+    }
+
+    pub fn soname(&self) -> Result<String, Error> {
+        let section_headers = self.read_section_headers()?;
```

Comment:
For some reason this call is always failing when encountering our executables in Firefox, even though it works fine for system libraries. Could you check with one of our builds and see what's going on?

Note that I tried this with an elf-hack build because that's what we ship.

Reply:
Interesting, I'll try it out!

Reply:
I'm unable to reproduce this, both with firefox release and nightly (as downloaded directly from our website). Is there something else I need to do? Can you tell me exactly what you did to test? I added a test like
```rust
    #[test]
    fn soname_firefox() {
        let bytes = include_bytes!("/home/afranchuk/Downloads/125/firefox/libxul.so");
        let reader = ModuleReader::new(&bytes[..]).unwrap();
        let soname = reader.soname().unwrap();
        assert_eq!(soname, "libxul.so");
    }
```

I tried with a few other libraries too.

Reply:
I patched the gecko sources to build minidump-writer with this PR included and crashed a child process to check how the code would behave in the wild.

Reply:
Perhaps this is a vaddr vs file offset issue. I'll try that.